### PR TITLE
Issue 4641 - IFC GUID and Revit ID support in v4 groups

### LINKED ISSUE #4641: v4 Groups - full IFC GUID and revit ID support
### Description
We're introducing Revit ID support and also better IFC Guid support in https://github.com/3drepo/3drepo.io/issues/4552 for ticket groups.

We also need this on v4 groups.

### Goals
- [ ] As a user I want my groups to work across revisions if I am providing a BIM model that has persistent IFC guids or Revit IDs

### Tasks
TBC

### PR DESCRIPTION
**Pull request by carmenfan**
This fixes #4641

#### Description
- Add generic IFC GUIDs and REVIT support to v4 groups
- refactored v5 code to expose some functions to be used by v4

#### Test cases
- v5 groups should function as before
- functionality added in #4552 for v5 groups should also work for v4 groups

### CODE REVIEW

File: backend/src/v4/models/group.js [hunk 1]
```diff
@@ -91,15 +101,22 @@ function clean(groupData) {
 	cleanArray(groupData, "transformation");
 
 	for (let i = 0; groupData.objects && i < groupData.objects.length; i++) {
-		cleanArray(groupData.objects[i], "ifc_guids");
+		Object.values(idTypes).forEach((idType) => {
+			if (groupData.objects[i][idType]) {
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v4/models/group.js:105]: we could, but I don't think it really gives anything here.

we have to do a loop to get the current elements, then we have to do a loop (whilst it should only ever have 1 result, it should still be a loop or it's not very future proof and also makes it difficult to read)

so O(2n) when i'm doing it in O(n) right now. I don't think it improves readability either.

File: backend/src/v4/models/group.js [hunk 2]
```diff
@@ -111,146 +128,122 @@ function getGroupCollectionName(model) {
...
-	for (let i = 0; i < groupData.objects.length; i++) {
-		const objectId = {};
-		objectId.account = groupData.objects[i].account;
-		objectId.model = groupData.objects[i].model;
+	return Promise.all(groupData.objects.map(async({account, model:container, shared_ids, ...extIds}) => {
 
-		const _branch = (model === objectId.model) ? branch : "master";
-		const _revId = (model === objectId.model) ? revId : null;
+		if (showIfcGuids) {
+			if (extIds[idTypes.IFC]) {
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v4/models/group.js:145]: the v4 schema doesn't support rvt. we're not adding new features in v4 to expose it (as of now)

So rvt is only used internally for version control benefits

File: backend/src/v4/models/group.js [hunk 3]
```diff
@@ -111,146 +128,122 @@ function getGroupCollectionName(model) {
...
+				return {account, model: container, [idTypes.IFC]: extIds[idTypes.IFC]};
+			}
+			return {account, model: container, [idTypes.IFC]: []};
+		} else if(shared_ids) {
+			return {account, model: container, shared_ids: convertSharedIDsToString ? shared_ids.map(utils.uuidToString) : shared_ids.map(utils.stringToUUID)};
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v4/models/group.js:150]: welcome to v4! Unlike v5 where we enforce conversion in middleware, in v4, you can never be sure. 😢

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v4/models/group.js:150]: this function is both used by reading and writing from database, the reading bit would come in with uuids, the writing bit would come in as strings...

File: backend/src/v4/models/group.js [hunk 4]
```diff
@@ -111,146 +128,122 @@ function getGroupCollectionName(model) {
...
+		// At this point, we're either trying to convert shared Ids to ifcGuids, or external ids to shared ids
+		const {_id: conRevId} = await getHistory(
+			account, container,
+			model === container ? branch : "master",
+			model === container && !branch ? revId : undefined,
+			{_id: 1}
+		);
 
-		const objectIdsSet = groupData.objects[i].shared_ids ? new Set(groupData.objects[i].shared_ids.map(utils.uuidToString)) : new Set();
+		if(shared_ids) {
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v4/models/group.js:161]: good point, the earlier if statement shouldn't return if there's no Ifc_guids so the bug is up there not here.

File: backend/src/v4/models/group.js [hunk 5]
```diff
@@ -323,7 +314,7 @@ Group.findByUID = async function (account, model, branch, revId, uid, showIfcGui
 	}
 
 	if (convertToIfcGuids) {
```
Comment (ChristosTsiotsias) [inline backend/src/v4/models/group.js:316]: Im confused as to why we have both showIfcGuids and convertToIfcGuids. Isnt getObjectsArrayAsExternalIds doing the same as getObjectIds with showIfcGuids set to true ?

**Review by ChristosTsiotsias — CHANGES_REQUESTED** [review]

File: backend/src/v4/models/group.js [hunk 6]
```diff
@@ -111,146 +131,118 @@ function getGroupCollectionName(model) {
...
+		// At this point, we're either trying to convert shared Ids to ifcGuids, or external ids to shared ids
+		const {_id: conRevId} = await getHistory(
+			account, container,
+			model === container ? branch : "master",
+			model === container && !branch ? revId : undefined,
+			{_id: 1}
+		);
 
-		const objectIdsSet = groupData.objects[i].shared_ids ? new Set(groupData.objects[i].shared_ids.map(utils.uuidToString)) : new Set();
+		if(shared_ids) {
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v4/models/group.js:163]: not sure i'd agree, this chunk are early rejects

image


If we don't do early rejects then yes i agree but i'd also have to rewrite it for a different flow.

This is v4 so i'm doing minimal changes to keep it working really.

File: backend/src/v4/models/group.js [hunk 7]
```diff
@@ -394,7 +385,7 @@ Group.getList = async function (account, model, branch, revId, ids, queryParams,
 Group.update = async function (account, model, branch = "master", revId = null, sessionId, user = "", groupId, data) {
 	const group = await Group.findByUID(account, model, branch, revId, groupId);
 
-	const convertedObjects = await getObjectsArrayAsIfcGuids(data, false);
+	const convertedObjects = Array.isArray(data.objects) ? await getObjectsArrayAsExternalIds(account, model, branch, revId, data, false) : undefined;
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v4/models/group.js:388]: yes you're correct, but at L445 it's generic so the update could've been external IDs so the original group data doesn't guarantee to be shared ids

(yes we could add some if statements etc, but trying to do minimal change to keep it working 😆 )

File: backend/src/v4/models/group.js (hunk 5 above)
Comment (carmenfan) [reply backend/src/v4/models/group.js:316]: Not clear to me either, If i were to hazard a guess convertToIfcGuids makes an assumption that we have shared Ids and we're trying to convert to externals, and `showIfcGuids` implies we're turning it into Ifc guids no matter what.

We could change it, but in the interest of not breaking something i don't understand (and there's very little tests to check for regression), i'd rather not touch it.

**Pull request merged**